### PR TITLE
Tw refactor.3

### PR DESCRIPTION
**Pull request by siosonel**
## Description

Updated [tw/README.md](https://github.com/stjude/proteinpaint/blob/tw-refactor.3/client/tw/README.md). Now recommends the `addons` option  to add custom methods to a tw subclass, because:
- extending every relevant subclass with custom plot/app methods may be unnecessarily repetitive 
- extending a custom composite base class will lose inheritance of specialized subclass methods that are specific to a term type + q.type/mode

Tested with
- http://localhost:3000/testrun.html?name=*.unit
- http://localhost:3000/testrun.html?name=*.integration
- `cd client && npm run tsc`

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

**Review by airenzp — DISMISSED** [review]
Looks great thanks

**Review by airenzp — APPROVED** [review]

File: client/tw/categorical.ts [hunk 1]
```diff
@@ -139,8 +123,8 @@ export class CatValues extends CategoricalBase {
 
 		// GDC or other dataset may allow missing term.values
 		if (!term.values) term.values = {}
-		const numVals = Object.keys(tw.term.values).length
-		// GDC or other dataset may allow empty term.values
+		//const numVals = Object.keys(tw.term.values).length
+		//GDC or other dataset may allow empty term.values
 		//if (!numVals) throw `empty term.values`
 		if (q.mode == 'binary') {
```
Comment (gavrielm) [inline client/tw/categorical.ts:129]: Should not allow `mode='binary'` here. Should only be allowed for custom groupsetting (for categorical term).

When verifying `mode='binary'`, should check `tw.q.customset`, not `tw.term.values`.

Can address in a later PR if you prefer.

**Comment by gavrielm** [comment]
Maybe related to comment above, not sure why the `mode` property is necessary in these term definitions:

```
export type ValuesQ = MinBaseQ & {
	type: 'values'
	mode?: 'binary'
}

export type PredefinedGroupSettingQ = MinBaseQ & {
	type: 'predefined-groupset'
	predefined_groupset_idx: number
	mode?: 'binary'
}

export type CustomGroupSettingQ = MinBaseQ & {
	type: 'custom-groupset'
	customset: BaseGroupSet
	mode?: 'binary'
}
```

File: client/tw/categorical.ts [hunk 2]
```diff
@@ -162,7 +146,7 @@ export class CatValues extends CategoricalBase {
 }
 
 export class CatPredefinedGS extends CategoricalBase {
-	//term: CategoricalTerm
+	// term, type, isAtomic, $id are set in ancestor base classes
 	q: PredefinedGroupSettingQ
 	#groupset: BaseGroupSet
```
Comment (gavrielm) [inline client/tw/categorical.ts:151]: Why is `#groupset: BaseGroupSet` necessary? Term groupsetting should already be in `this.term`.

**Comment by gavrielm** [comment]
In `tw/numeric.ts`:

```
if (preferredBins) {
	const q = opts.defaultQ
	if (tw.term.bins[preferredBins]) {
		tw.q = structuredClone(tw.term.bins[preferredBins])
		delete q.preferredBins
	} else if (preferredBins == 'median') {
		if (q.type != 'custom-bin') throw '.type must be custom-bin when .preferredBins=median'
		Object.assign(tw.q, q)
	} else {
		throw `unrecognized defaultQ.preferredBins='${preferredBins}'`
	}
}
```

It looks like if `tw.term.bins[preferredBins]`, then `defaultQ` will not be merged into `q`

**Comment by gavrielm** [comment]
Hard to follow this code block in `tw/numeric.ts`:

```
if (tw.term.bins) {
	// detect if tw.q still needs to use term.bins
	//
	// TODO: may move some of these logic in the applicable switch-case code block,
	//       which may require not allowwing "mixed-modes", like some code expecting
	//       q.type='regular-bin' to possibly have q.mode='binary' | 'continuous',
	//       should clean or clarify such usecases
	//
	...
```
Will be good to do the TODO mentioned in the comment to help clarify the logic.

File: client/tw/categorical.ts (hunk 2 above)
Comment (siosonel) [reply client/tw/categorical.ts:151]: setting a `groupset` instance property would allow abstraction of what groupset refers to, between predefined-groupset and custom-groupset. In this way, the same addon method can refer to just `this.groupset` without needing to test if it's a CatPredefinedGS or CatCustomGS. However, in order for an addon method to access the `groupset` property, it cannot be private (with a hash prefix), I'll need to change to a non-enumerable `groupset` property (I'll do it in this PR).

**Comment by siosonel** [comment]
> In `tw/numeric.ts`:
> 
> ```
> if (preferredBins) {
> 	const q = opts.defaultQ
> 	if (tw.term.bins[preferredBins]) {
> 		tw.q = structuredClone(tw.term.bins[preferredBins])
> 		delete q.preferredBins
> 	} else if (preferredBins == 'median') {
> 		if (q.type != 'custom-bin') throw '.type must be custom-bin when .preferredBins=median'
> 		Object.assign(tw.q, q)
> 	} else {
> 		throw `unrecognized defaultQ.preferredBins='${preferredBins}'`
> 	}
> }
> ```
> 
> It looks like if `tw.term.bins[preferredBins]`, then `defaultQ` will not be merged into `q`

Yes, the code assumes that `defaultQ.preferredBins` indicates to only consider `tw.term.bins[preferredBins]`, that `defaultQ` has no other relevant properties to merge into `tw.q`. But if that is a wrong assumption, should `opts.defaultQ` **always** be merged to `tw.q`, before doing anything else? But that would mean that the `tw.term.bins.default | less` that's merged later would replace matching property keys in "preferred bins". Or maybe, set tw.q to the preferred bins first, then apply the defaultQ overrides? So, would need to resolve the order in which to apply the overrides versus preferred bins.

**Review by gavrielm — APPROVED** [review]
